File: ts-sdk/client/src/txbuilder/openTunaSpotPosition.ts
import { FUSIONAMM_PROGRAM_ADDRESS, FusionPool, getFusionPoolDecoder } from "@crypticdot/fusionamm-client";
import { getWhirlpoolDecoder, Whirlpool, WHIRLPOOL_PROGRAM_ADDRESS } from "@orca-so/whirlpools-client";
import {
  type Account,
  Address,
  decodeAccount,
  fetchEncodedAccount,
  GetAccountInfoApi,
  GetMultipleAccountsApi,
  IInstruction,
  Rpc,
  TransactionSigner,
} from "@solana/kit";
import {
  ASSOCIATED_TOKEN_PROGRAM_ADDRESS,
  fetchAllMaybeMint,
  findAssociatedTokenPda,
  Mint,
} from "@solana-program/token-2022";
import assert from "assert";

import {
  getMarketAddress,
  getOpenTunaSpotPositionInstruction,
  getTunaSpotPositionAddress,
  OpenTunaSpotPositionInstructionDataArgs,
} from "../index.ts";

export async function openTunaSpotPositionInstructions(
  rpc: Rpc<GetAccountInfoApi & GetMultipleAccountsApi>,
  authority: TransactionSigner,
  pool: Address,
  args: OpenTunaSpotPositionInstructionDataArgs,
): Promise<IInstruction[]> {
  const encodedPool = await fetchEncodedAccount(rpc, pool);
  assert(encodedPool.exists, "Pool account not found");

  let poolAccount: Account<FusionPool> | Account<Whirlpool>;
  if (encodedPool.programAddress == FUSIONAMM_PROGRAM_ADDRESS) {
    poolAccount = decodeAccount(encodedPool, getFusionPoolDecoder());
  } else if (encodedPool.programAddress == WHIRLPOOL_PROGRAM_ADDRESS) {
    poolAccount = decodeAccount(encodedPool, getWhirlpoolDecoder());
  } else {
    throw new Error("Incorrect Fusion or Orca pool account");
  }

  const [mintA, mintB] = await fetchAllMaybeMint(rpc, [poolAccount.data.tokenMintA, poolAccount.data.tokenMintB]);
  assert(mintA.exists, "Token A account not found");
  assert(mintB.exists, "Token B account not found");

  const ix = await openTunaSpotPositionInstruction(authority, mintA, mintB, pool, args);
  return [ix];
}

export async function openTunaSpotPositionInstruction(
  authority: TransactionSigner,
  mintA: Account<Mint>,
  mintB: Account<Mint>,
  pool: Address,
  args: OpenTunaSpotPositionInstructionDataArgs,
): Promise<IInstruction> {
  const tunaPositionAddress = (await getTunaSpotPositionAddress(authority.address, pool))[0];
  const marketAddress = (await getMarketAddress(pool))[0];

  const tunaPositionAtaA = (
    await findAssociatedTokenPda({
      owner: tunaPositionAddress,
      mint: mintA.address,
      tokenProgram: mintA.programAddress,
    })
  )[0];

  const tunaPositionAtaB = (
    await findAssociatedTokenPda({
      owner: tunaPositionAddress,
      mint: mintB.address,
      tokenProgram: mintB.programAddress,
    })
  )[0];

  return getOpenTunaSpotPositionInstruction({
    authority,
    mintA: mintA.address,
    mintB: mintB.address,
    tokenProgramA: mintA.programAddress,
    tokenProgramB: mintB.programAddress,
    tunaPosition: tunaPositionAddress,
    tunaPositionAtaA,
    tunaPositionAtaB,
    pool: pool,
    market: marketAddress,
    associatedTokenProgram: ASSOCIATED_TOKEN_PROGRAM_ADDRESS,
    ...args,
  });
}

File: rust-sdk/client/src/txbuilder/open_tuna_spot_position.rs
use crate::instructions::{OpenTunaSpotPosition, OpenTunaSpotPositionInstructionArgs};
use crate::types::PoolToken;
use crate::{get_market_address, get_tuna_spot_position_address};
use anyhow::{anyhow, Result};
use fusionamm_client::{FusionPool, FUSIONAMM_ID};
use orca_whirlpools_client::{Whirlpool, WHIRLPOOL_ID};
use solana_client::rpc_client::RpcClient;
use solana_instruction::Instruction;
use solana_pubkey::Pubkey;
use solana_sdk_ids::system_program;
use spl_associated_token_account::get_associated_token_address_with_program_id;

impl Default for OpenTunaSpotPositionInstructionArgs {
    fn default() -> Self {
        OpenTunaSpotPositionInstructionArgs {
            position_token: PoolToken::A,
            collateral_token: PoolToken::A,
        }
    }
}

pub fn open_tuna_spot_position_instructions(
    rpc: &RpcClient,
    authority: &Pubkey,
    pool: &Pubkey,
    args: OpenTunaSpotPositionInstructionArgs,
) -> Result<Vec<Instruction>> {
    let pool_account = rpc.get_account(pool)?;

    let (mint_a_address, mint_b_address) = if pool_account.owner == FUSIONAMM_ID {
        let pool = FusionPool::from_bytes(&pool_account.data)?;
        (pool.token_mint_a, pool.token_mint_b)
    } else if pool_account.owner == WHIRLPOOL_ID {
        let pool = Whirlpool::from_bytes(&pool_account.data)?;
        (pool.token_mint_a, pool.token_mint_b)
    } else {
        return Err(anyhow!("Incorrect fusion or orca pool"));
    };

    let mint_accounts = rpc.get_multiple_accounts(&[mint_a_address.into(), mint_b_address.into()])?;
    let mint_a_account = mint_accounts[0].as_ref().ok_or(anyhow!("Token A mint account not found"))?;
    let mint_b_account = mint_accounts[1].as_ref().ok_or(anyhow!("Token B mint account not found"))?;

    let mut instructions = vec![];

    instructions.push(open_tuna_spot_position_instruction(
        authority,
        pool,
        &mint_a_address,
        &mint_b_address,
        &mint_a_account.owner,
        &mint_b_account.owner,
        args,
    ));

    Ok(instructions)
}

pub fn open_tuna_spot_position_instruction(
    authority: &Pubkey,
    pool: &Pubkey,
    mint_a: &Pubkey,
    mint_b: &Pubkey,
    token_program_a: &Pubkey,
    token_program_b: &Pubkey,
    args: OpenTunaSpotPositionInstructionArgs,
) -> Instruction {
    let tuna_position_address = get_tuna_spot_position_address(authority, pool).0;
    let market = get_market_address(pool).0;

    let ix_builder = OpenTunaSpotPosition {
        authority: *authority,
        mint_a: *mint_a,
        mint_b: *mint_b,
        token_program_a: *token_program_a,
        token_program_b: *token_program_b,
        market,
        tuna_position: tuna_position_address,
        tuna_position_ata_a: get_associated_token_address_with_program_id(&tuna_position_address, mint_a, token_program_a),
        tuna_position_ata_b: get_associated_token_address_with_program_id(&tuna_position_address, mint_b, token_program_b),
        pool: *pool,
        system_program: system_program::ID,
        associated_token_program: spl_associated_token_account::ID,
    };

    ix_builder.instruction(args)
}
